# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/kishan992/FlyRank-ML-Internship/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*

**Lane Name:** Refresh / Content Opportunity Scoring

**Our Question:** Can we use the last 90 days of search numbers (like page views, positions, and user scroll rates) to rank our web pages and find the ones that are losing traffic the fastest?

**Why this lane:** Big websites have hundreds of thousands of pages, but a small team can only fix a few pages a week. If they just guess which pages to update, they waste time. This lane lets us build a smart sorting system that hands the team a top-priority list of exactly which dying pages they should fix first to save traffic and money.

In [6]:
import pandas as pd

# Pulling the file directly from my GitHub repository raw URL
url = "https://raw.githubusercontent.com/kishan992/FlyRank-ML-Internship/main/data/raw/content_refresh_anonymized.csv"

try:
    df_starter = pd.read_csv(url)
    print("Starter slice loaded successfully via GitHub Raw URL!")
    print(f"Total rows in this local slice: {len(df_starter):,}")
    print(f"Sample columns available for our model: {list(df_starter.columns[:6])}")
except Exception as e:
    print("Error loading the file:")
    print(e)

Starter slice loaded successfully via GitHub Raw URL!
Total rows in this local slice: 30,000
Sample columns available for our model: ['content_id', 'client_id', 'search_volume', 'competition', 'competition_level', 'cpc']


## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*

**What decision does your work improve?**
It helps us decide exactly which web pages our writing team should spend their limited time updating each week.

**Who acts on it?**
A Content Operations Manager or Editorial Team Lead will act on it by taking our top-ranked list and assigning those pages to writers.

**What does a wrong recommendation cost?**
* **If we flag a healthy page by mistake (False Positive):** We waste valuable budget and hours paying a writer to rewrite a page that didn't need any work.
* **If we miss a dying page by mistake (False Negative):** High-value pages quietly rot and lose their ranks, causing the company to lose search traffic and revenue to competitors.

In [7]:
# Let's see the average value (Cost Per Click) of the pages we are monitoring (This shows what a wrong recommendation could cost)
average_cpc = df_starter['cpc'].mean()
highest_cpc = df_starter['cpc'].max()

print("Data Check for Section 2:")
print(f"Average Cost Per Click (CPC) across these pages: ${average_cpc:.2f}")
print(f"Highest single Cost Per Click in this dataset: ${highest_cpc:.2f}")

Data Check for Section 2:
Average Cost Per Click (CPC) across these pages: $0.49
Highest single Cost Per Click in this dataset: $100.36


## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

**Key Numbers from our Dataset:**
1. **Total Scale:** We are monitoring a starter pool of 30,000 unique web pages.
2. **Financial Value:** The average Cost Per Click is 0.49 dollars, with a massive maximum cap of 100.36 dollars, proving these pages hold serious financial weight.
3. **Competitive Landscape:** There are 2,658 high-competition pages at extreme risk of losing traffic to competitors, and 1,836 medium-competition pages ripe for a ranking boost.

In [8]:
# Let's count how many pages fall into high competition tiers (This data shows the status of pages based on their traffic level)
competition_counts = df_starter['competition_level'].value_counts()

print("Data Check for Section 3:")
print("Breakdown of pages by competition level:")
print(competition_counts)


Data Check for Section 3:
Breakdown of pages by competition level:
competition_level
LOW       22896
HIGH       2658
MEDIUM     1836
Name: count, dtype: int64


## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

**What this model CAN do (Our Claims):**
* **Provide Decision Support:** It will rank our web pages so the editorial team knows exactly which articles to update first.
* **Identify Directional Trends:** It will spot which pages are showing patterns of fading traffic, decreasing impressions, or dropping ranks.
* **Surface Hidden Risks:** It will flag high-value pages (like those with high Cost Per Click) that are quietly losing steam before the traffic drop hurts company revenue.

**What this model NEVER will do (Our Boundaries):**
* **Claim Causal Proof:** It cannot prove *why* traffic dropped (e.g., it can't tell if a drop was caused by a Google algorithm change, a competitor's new backlink, or seasonal trends). It only flags *that* a drop is happening.
* **"Predict" Google:** We are not engineering a tool that reads Google's mind or forecasts future search engine updates. We are simply building an early-warning system based on historical performance data.

In [9]:
# No code for Section-4


## Self-check

Before you submit, confirm each line honestly:

- [x] Every section above is filled — markdown thinking AND the code that backs it
- [x] The notebook runs top to bottom with no errors (Runtime → Run all)
- [x] No client names, URLs, or private queries anywhere
- [x] My claims use careful words: observed, measured, directional, decision-support
- [x] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.